### ISSUE TITLE
C unions of structs with different sizes are corrupted while passed to Crystal

### ISSUE THREAD

**Issue by Hadeweka** [crystal-lang/crystal#12550]
## Bug Report

Tested on Crystal 1.6.0-dev version for Windows (from 2022-09-15, bab729a), but seems to be reproducible on other platforms and versions as well.

The following C code simply passes a chunk of data with total size of 8 bytes:
```c
typedef struct LargeStruct {
  char padding[8];
} LargeStruct;

extern void set_chunk(LargeStruct* val) {
  for(int i = 0; i < 8; i++) {
    val->padding[i] = (char) i;
  }
}
```

Then, if the following Crystal binding is used:
```crystal
@[Link(ldflags: "large_struct.obj")]
lib FaultyLib
  union PassedValue
    faulty : FaultyStruct
    other : OtherStruct
    padding : UInt8[8]
  end

  struct FaultyStruct
    value_32 : UInt32
    value_8_1 : UInt8
    value_8_2 : UInt8
    value_8_3 : UInt8
    value_8_4 : UInt8
  end

  struct OtherStruct
    value_32 : UInt32
    value_64 : UInt64
  end

  fun set_chunk(uni : PassedValue*)
end

def inner_func
  FaultyLib.set_chunk(out dummy)
  puts "Inner: #{dummy}"
  puts "Inner inspect: #{dummy.inspect}"
  dummy
end

def outer_func
  puts "Outer inspect: #{inner_func.inspect}"
end

outer_func
```
the following output is generated:
```
Inner:         FaultyLib::PassedValue(@faulty=FaultyLib::FaultyStruct(@value_32=50462976, @value_8_1=246, @value_8_2=127, @value_8_3=0, @value_8_4=0), @other=FaultyLib::OtherStruct(@value_32=50462976, @value_64=2685284929088), @padding=StaticArray[0, 1, 2, 3, 246, 127, 0, 0])
Inner inspect: FaultyLib::PassedValue(@faulty=FaultyLib::FaultyStruct(@value_32=50462976, @value_8_1=4, @value_8_2=5, @value_8_3=6, @value_8_4=7), @other=FaultyLib::OtherStruct(@value_32=50462976, @value_64=2685284929088), @padding=StaticArray[0, 1, 2, 3, 4, 5, 6, 7])
Outer inspect: FaultyLib::PassedValue(@faulty=FaultyLib::FaultyStruct(@value_32=50462976, @value_8_1=0, @value_8_2=0, @value_8_3=0, @value_8_4=0), @other=FaultyLib::OtherStruct(@value_32=50462976, @value_64=2685284929088), @padding=StaticArray[0, 1, 2, 3, 0, 0, 0, 0])
```

The second line (generated by `dummy.inspect`) is the expected behavior, while the other lines are all corrupted, but only after the first 4 bytes. This indicates that some degree of corruption is already present as soon as the intial value is obtained from the C lib function.

Removing `other` from the union will lead to expected behavior in all three outputs again, but removing `padding` from the union will not fix the problem.

**Comment by HertzDevil** [comment]
It seems this has nothing to do with lib funs at all:

```crystal
@[Extern]
record Foo, x : UInt8, y : UInt16

@[Extern(union: true)]
record Bar, foo : Foo, padding : UInt8[4]

def inner_func
  dummy = uninitialized Bar
  dummy.padding.fill { |i| 100_u8 + i }
  puts "Inner: #{dummy}"
  puts "Inner inspect: #{dummy.inspect}"
  dummy
end

def outer_func
  puts "Outer inspect: #{inner_func.inspect}"
end

outer_func
```

```
Inner: Bar(@foo=Foo(@x=100, @y=26470), @padding=StaticArray[100, 0, 102, 103])
Inner inspect: Bar(@foo=Foo(@x=100, @y=26470), @padding=StaticArray[100, 101, 102, 103])
Outer inspect: Bar(@foo=Foo(@x=100, @y=26470), @padding=StaticArray[100, 0, 102, 103])
```

It probably has something to do with the member padding inside `OtherStruct` and `Foo`.

**Comment by straight-shoota** [comment]
Not sure if this can be a padding issue. `StaticArray` should have no padding between its members 😅 

Found another variant:
```cr
puts "Inner tos: #{tos(dummy)}" # ... @padding=StaticArray[100, 223, 102, 103]

def tos(dummy)
  dummy.to_s
end
```
And the number at `@padding[1]` varies between runs, even with the same binary. 223 was relatively frequent, but I also got 111, 47, 15 and other numbers.
Looks almost as if it was hooked up to a RNG (with some bias, though).

**Comment by HertzDevil** [comment]
By "member padding" I mean the bytes between `OtherStruct#@value_32` and `@value_64`, or between `Foo#@x` and `@y`. It seems they are the parts that are lost.

**Comment by asterite** [comment]
I have a sense that this one will be pretty tricky to fix.

If you have code like this in C:

```crystal
typedef struct Foo {
  char x;
  short y;
} Foo;

typedef union Bar {
  Foo foo;
  char padding[4];
} Bar;

extern void set_union(Bar bar) {
  bar.padding[0] = 1;
  bar.padding[1] = 2;
  bar.padding[2] = 3;
  bar.padding[3] = 4;
}
```

This is the LLVM IR code clang generates:

```llvm
%union.Bar = type { %struct.Foo }
%struct.Foo = type { i8, i16 }

; Function Attrs: noinline nounwind optnone ssp uwtable
define void @set_union(i64 %0) #0 {
  %2 = alloca %union.Bar, align 2
  %3 = alloca i64, align 8
  %4 = getelementptr inbounds %union.Bar, %union.Bar* %2, i32 0, i32 0
  store i64 %0, i64* %3, align 8
  %5 = bitcast %struct.Foo* %4 to i8*
  %6 = bitcast i64* %3 to i8*
  call void @llvm.memcpy.p0i8.p0i8.i64(i8* align 2 %5, i8* align 8 %6, i64 4, i1 false)
  %7 = bitcast %union.Bar* %2 to [4 x i8]*
  %8 = getelementptr inbounds [4 x i8], [4 x i8]* %7, i64 0, i64 0
  store i8 1, i8* %8, align 2
  %9 = bitcast %union.Bar* %2 to [4 x i8]*
  %10 = getelementptr inbounds [4 x i8], [4 x i8]* %9, i64 0, i64 1
  store i8 2, i8* %10, align 1
  %11 = bitcast %union.Bar* %2 to [4 x i8]*
  %12 = getelementptr inbounds [4 x i8], [4 x i8]* %11, i64 0, i64 2
  store i8 3, i8* %12, align 2
  %13 = bitcast %union.Bar* %2 to [4 x i8]*
  %14 = getelementptr inbounds [4 x i8], [4 x i8]* %13, i64 0, i64 3
  store i8 4, i8* %14, align 1
  ret void
}
```

Note that the alignments of the assignments to `padding` are 2, 1, 2 and 1 respectively. 

If we change `short y` to `int y`, we get this instead:

```llvm
%union.Bar = type { %struct.Foo }
%struct.Foo = type { i8, i32 }

; Function Attrs: noinline nounwind optnone ssp uwtable
define void @set_union(i64 %0) #0 {
  %2 = alloca %union.Bar, align 4
  %3 = getelementptr inbounds %union.Bar, %union.Bar* %2, i32 0, i32 0
  %4 = bitcast %struct.Foo* %3 to i64*
  store i64 %0, i64* %4, align 4
  %5 = bitcast %union.Bar* %2 to [4 x i8]*
  %6 = getelementptr inbounds [4 x i8], [4 x i8]* %5, i64 0, i64 0
  store i8 1, i8* %6, align 4
  %7 = bitcast %union.Bar* %2 to [4 x i8]*
  %8 = getelementptr inbounds [4 x i8], [4 x i8]* %7, i64 0, i64 1
  store i8 2, i8* %8, align 1
  %9 = bitcast %union.Bar* %2 to [4 x i8]*
  %10 = getelementptr inbounds [4 x i8], [4 x i8]* %9, i64 0, i64 2
  store i8 3, i8* %10, align 2
  %11 = bitcast %union.Bar* %2 to [4 x i8]*
  %12 = getelementptr inbounds [4 x i8], [4 x i8]* %11, i64 0, i64 3
  store i8 4, i8* %12, align 1
  ret void
}
```

Note the store alignments are 4, 1, 2 and 1.

Something interesting I found...

If we have `cool_lib.c` to be this:

```c
typedef struct Foo {
  char x;
  int y;
} Foo;

typedef union Bar {
  Foo foo;
  char padding[4];
} Bar;

extern void set_union(Bar* bar) {
  bar->padding[0] = 1;
  bar->padding[1] = 2;
  bar->padding[2] = 3;
  bar->padding[3] = 4;
}
```

And the Crystal program:

```crystal
@[Link(ldflags: "#{__DIR__}/cool_lib.o")]
lib CoolLib
  struct Foo
    x : UInt8
    y : Int32
  end

  union Bar
    foo : Foo
    padding : UInt8[4]
  end

  fun set_union(uni : Bar*)
end

dummy = uninitialized CoolLib::Bar
CoolLib.set_union(pointerof(dummy))
puts dummy
puts dummy.padding
puts dummy.unsafe_as(StaticArray(UInt8, 4))
ptr = dummy.padding.to_unsafe
puts ptr[0]
puts ptr[1]
puts ptr[2]
puts ptr[3]
```

Then the output is:

```
CoolLib::Bar(@foo=CoolLib::Foo(@x=1, @y=0), @padding=StaticArray[1, 207, 184, 0])
StaticArray[1, 2, 3, 4]
StaticArray[1, 2, 3, 4]
1
2
3
4
```

It seems the correct data is there. It's just printed wrong when inspecting the union... for some reason.

Similarly, if we add `puts "Inner: #{dummy.padding}"` to the `inner_func` of the first example, it looks like `padding` has actually the correct values. But then when the union is returned to `outer_func` the value gets corrupted for some reason:

```crystal
@[Link(ldflags: "#{__DIR__}/cool_lib.o")]
lib FaultyLib
  union PassedValue
    faulty : FaultyStruct
    other : OtherStruct
    padding : UInt8[8]
  end

  struct FaultyStruct
    value_32 : UInt32
    value_8_1 : UInt8
    value_8_2 : UInt8
    value_8_3 : UInt8
    value_8_4 : UInt8
  end

  struct OtherStruct
    value_32 : UInt32
    value_64 : UInt64
  end

  fun set_chunk(uni : PassedValue*)
end

def inner_func
  dummy = uninitialized FaultyLib::PassedValue
  FaultyLib.set_chunk(pointerof(dummy))
  puts "Inner: #{dummy}"
  puts "Inner: #{dummy.padding}"
  puts "Inner inspect: #{dummy.inspect}"
  dummy
end

def outer_func
  dummy = inner_func
  puts "Outer: #{dummy}"
  puts "Outer: #{dummy.padding}"
  puts "Outer inspect: #{dummy.inspect}"
end

outer_func
```

Output:

```
Inner: FaultyLib::PassedValue(@faulty=FaultyLib::FaultyStruct(@value_32=50462976, @value_8_1=1, @value_8_2=0, @value_8_3=0, @value_8_4=0), @other=FaultyLib::OtherStruct(@value_32=50462976, @value_64=0), @padding=StaticArray[0, 1, 2, 3, 1, 0, 0, 0])
Inner: StaticArray[0, 1, 2, 3, 4, 5, 6, 7]
Inner inspect: FaultyLib::PassedValue(@faulty=FaultyLib::FaultyStruct(@value_32=50462976, @value_8_1=4, @value_8_2=5, @value_8_3=6, @value_8_4=7), @other=FaultyLib::OtherStruct(@value_32=50462976, @value_64=0), @padding=StaticArray[0, 1, 2, 3, 4, 5, 6, 7])
Outer: FaultyLib::PassedValue(@faulty=FaultyLib::FaultyStruct(@value_32=50462976, @value_8_1=1, @value_8_2=0, @value_8_3=0, @value_8_4=0), @other=FaultyLib::OtherStruct(@value_32=50462976, @value_64=0), @padding=StaticArray[0, 1, 2, 3, 1, 0, 0, 0])
Outer: StaticArray[0, 1, 2, 3, 1, 0, 0, 0]
Outer inspect: FaultyLib::PassedValue(@faulty=FaultyLib::FaultyStruct(@value_32=50462976, @value_8_1=1, @value_8_2=0, @value_8_3=0, @value_8_4=0), @other=FaultyLib::OtherStruct(@value_32=50462976, @value_64=0), @padding=StaticArray[0, 1, 2, 3, 1, 0, 0, 0])
```

**Comment by asterite** [comment]
Here's another interesting discovery. Take a look at this:

```crystal
lib CoolLib
  struct Foo
    x : UInt8
    y : Int32
  end

  union Bar
    foo : Foo
    padding : UInt8[4]
  end
end

struct Struct
  def doit
    {% for ivar, i in @type.instance_vars %}
      puts @{{ivar.id}}
    {% end %}
  end
end

def doit2(dummy)
  dummy.doit
end

dummy = uninitialized CoolLib::Bar
dummy.padding.fill { |i| 1_u8 + i }
print "puts dummy:"
puts dummy
print "puts dummy.padding:"
puts dummy.padding
puts "dummy.doit:"
dummy.doit
puts "doit2(dummy)"
doit2(dummy)
```

The output is:

```crystal
puts dummy:CoolLib::Bar(@foo=CoolLib::Foo(@x=1, @y=0), @padding=StaticArray[1, 143, 158, 0])
puts dummy.padding:StaticArray[1, 2, 3, 4]
dummy.doit:
CoolLib::Foo(@x=1, @y=0)
StaticArray[1, 2, 3, 4]
doit2(dummy)
CoolLib::Foo(@x=1, @y=0)
StaticArray[1, 230, 78, 0]
```

So now I think the issue is that C union types are not being passed around correctly. Note that when we call `doit` right on the stack frame that has the `dummy` variable, it works fine. But when we pass it to another function and do the same thing, that's when it breaks.

**Comment by straight-shoota** [comment]
If you append another round of doit2, I get a different value for the second item (like in https://github.com/crystal-lang/crystal/issues/12550#issuecomment-1263823783).
```cr
# ...
puts "doit2(dummy)"
doit2(dummy)
puts "doit2(dummy) cont."
doit2(dummy)
```
```
...
doit2(dummy)
CoolLib::Foo(@x=1, @y=0)
StaticArray[1, 145, 99, 4]
doit2(dummy) cont.
CoolLib::Foo(@x=1, @y=0)
StaticArray[1, 146, 99, 4]
```
The second one is always one higher then the first. But it's different numbers every time. 😕

**Comment by asterite** [comment]
Clang seems to pass the union as an array of things, always. We are probably passing unions around the wrong way.

**Comment by HertzDevil** [comment]
If you have the following:

```crystal
lib Lib
  struct Foo
    x : UInt8
    y : UInt16
  end

  union Bar
    foo : Foo
    padding : UInt8[6] # larger than `Foo`
  end
end

a = uninitialized Lib::Bar
b = a
```

the generated LLVM IR is:

```llvm
%"union.Lib::Bar" = type { %"struct.Lib::Foo.wants_size", [2 x i8] }
%"struct.Lib::Foo.wants_size" = type { i8, i16 }

define %"union.Lib::Bar" @__crystal_main(i32 %argc, i8** %argv) {
alloca:
  %a = alloca %"union.Lib::Bar", align 8
  %b = alloca %"union.Lib::Bar", align 8
  br label %entry

entry:                                            ; preds = %alloca
  ; `Crystal::VERSION`...
  %0 = load %"union.Lib::Bar", %"union.Lib::Bar"* %a, align 2
  store %"union.Lib::Bar" %0, %"union.Lib::Bar"* %b, align 2
  %1 = load %"union.Lib::Bar", %"union.Lib::Bar"* %a, align 2
  ret %"union.Lib::Bar" %1
}
```

Judging from `Crystal::LLVMTyper#create_llvm_c_union_struct_type` it seems Crystal simply picks the union member with the largest alignment and then adds its own padding bytes only at the end.

**Comment by asterite** [comment]
> Loads and stores always use that union member, so any padding between that member's fields gets lost.

Where is that in the LLVM IR?

**Comment by HertzDevil** [comment]
from the definition of `%"union.Lib::Bar"`

**Comment by asterite** [comment]
But it seem that has the first member plus the padding, so why wouldn't a load and a store also include that?

**Comment by HertzDevil** [comment]
Again, the missing padding here refers to the byte between `%"struct.Lib::Foo.wants_size"`'s `i8` and `i16`, because that struct isn't packed.

**Comment by asterite** [comment]
Oh, thanks! I think I get it now. I'm surprised LLVM doesn't produce code to load and store that missing byte, given that the load and store refers to the entire thing.

**Comment by asterite** [comment]
What if instead of having the first union member be the type with the largest align, we have that be that but represented as an array of that alignment, to fill up that type's size? I'll send a PR and we can comment on that approach. So far it seems to be working. 

> On the other hand we cannot merely turn the union into a byte array because the array would effectively be overaligned

What would be the consequences of doing that?

**Comment by cyangle** [comment]
The interpreter also has issues with example code.

**Comment by asterite** [comment]
@cyangle Thanks! There are like 10 example codes here. Would you mind opening a separate issue for the interpreter with the one that fails? Just note that the interpreter can't and will likely never load `.o` or `.obj` files, so it's expected that the first example doesn't work.

The union stuff does work well, though.

**Comment by cyangle** [comment]
That's unfortunate. It means the interpreter doesn't work properly with shards loading obj files.

**Comment by straight-shoota** [comment]
@cyangle Interpreter behaviour is entirely different from this issue here. So I've created a new issue for that in #12554.
